R/S3.R

```diff
@@ -124,38 +124,94 @@ is_S3_class <- function(x) {
 }
 
 # -------------------------------------------------------------------------
-# Define a few base examples
+# Pull out validation functions so hit by code coverage
 
-S3_factor <- new_S3_class("factor",
-  function(.data = integer(), levels = character()) {
+validate_factor <- function(self) {
+  c(
+    if (typeof(self) != "integer")
+      "Underlying data must be an <integer>",
+    if (!is.character(attr(self, "levels")))
+      "attr(, 'levels') must be a <character>"
+  )
+}
+
+validate_date <- function(self) {
+  if (!is.numeric(self)) {
+    "Underlying data must be numeric"
+  }
+}
+
+validate_POSIXct <- function(self) {
+  if (!is.numeric(self)) {
+    return("Underlying data must be numeric")
+  }
+
+  tz <- attr(self, "tz")
+  if (!is.character(tz) || length(tz) != 1) {
+    return("attr(, 'tz') must be a single string")
+  }
```

Comment:
Are you supposed to `c()` the strings together like in `validate_factor()`?

Same question about `validate_data.frame()` I guess

Reply:
It's basically up to you and how much time you want to put into it. I don't think it's terrible if multiple problems need repeated calls to reveal every last detail.